uarray/_uarray_dispatch.cxx

```diff
@@ -179,14 +181,269 @@ std::string backend_to_domain_string(PyObject * backend) {
   return domain_to_string(domain.get());
 }
 
+struct BackendState {
+  PyObject_HEAD
+  global_state_t globals;
+  local_state_t locals;
+
+  static void dealloc(BackendState * self) {
+    self->~BackendState();
+    Py_TYPE(self)->tp_free(self);
+  }
+
+  static PyObject * new_(
+      PyTypeObject * type, PyObject * args, PyObject * kwargs) {
+    auto self = reinterpret_cast<BackendState *>(type->tp_alloc(type, 0));
+    if (self == nullptr)
+      return nullptr;
+
+    // Placement new
+    self = new (self) BackendState;
+    return reinterpret_cast<PyObject *>(self);
+  }
+
+  static PyObject * pickle_(BackendState * self) {
+    try {
+      py_ref py_global = BackendState::convert_py(self->globals);
+      py_ref py_locals = BackendState::convert_py(self->locals);
+
+      return Py_BuildValue("(OO)", py_global.get(), py_locals.get());
+    } catch (std::runtime_error &) {
+      return nullptr;
+    }
+  }
+
+  static PyObject * unpickle_(PyObject * cls, PyObject * args) {
+    try {
+      PyObject *py_locals, *py_global;
+      if (!PyArg_ParseTuple(args, "OO", &py_global, &py_locals))
+        return nullptr;
+      local_state_t locals = convert_local_state(py_ref::ref(py_locals));
+      global_state_t globals = convert_global_state(py_ref::ref(py_global));
+
+      py_ref empty_tuple = py_ref::steal(PyTuple_New(0));
+      if (!empty_tuple)
+        return nullptr;
+
+      BackendState * output = reinterpret_cast<BackendState *>(
+          PyObject_Call(cls, empty_tuple.get(), NULL));
+      if (output == nullptr)
+        return nullptr;
+
+      output->locals = locals;
+      output->globals = globals;
+
+      return reinterpret_cast<PyObject *>(output);
+    } catch (std::invalid_argument &) {
+      return nullptr;
+    } catch (std::bad_alloc &) {
+      PyErr_NoMemory();
+      return nullptr;
+    }
+  }
+
+  template <typename T, typename Convertor>
+  static std::vector<T> convert_iter(py_ref input, Convertor item_convertor) {
```

Comment:
None of these `convert_` functions need to manage ref-counts for their `input` argument so I would just pass them as `PyObject *`.

Reply:
We have `vector<py_ref>` in the outputs, which is why I chose `py_ref` here. Is there an easy way to convert that won't require a copy?